hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/connector/ReadJdbcP.java

```diff
@@ -0,0 +1,101 @@
+import com.hazelcast.jet.core.AbstractProcessor;
+import com.hazelcast.jet.core.ProcessorSupplier;
+import com.hazelcast.jet.core.processor.SourceProcessors;
+import com.hazelcast.jet.function.DistributedBiFunction;
+import com.hazelcast.jet.function.DistributedFunction;
+import com.hazelcast.jet.function.DistributedSupplier;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.Statement;
+
+import static com.hazelcast.jet.impl.util.Util.uncheckCall;
+
+/**
+ * Private API, use {@link SourceProcessors#readJdbcP}.
+ */
+public final class ReadJdbcP<T> extends AbstractProcessor {
+
+    private final DistributedSupplier<Connection> connectionSupplier;
+    private final DistributedFunction<Connection, Statement> statementFn;
+    private final DistributedBiFunction<Integer, Integer, String> sqlFn;
+    private final DistributedFunction<ResultSet, T> mapOutputFn;
+
+    private Connection connection;
+    private Statement statement;
+    private String sql;
+    private Traverser traverser;
+
+    private ReadJdbcP(
+            @Nonnull DistributedSupplier<Connection> connectionSupplier,
+            @Nonnull DistributedFunction<Connection, Statement> statementFn,
+            @Nonnull DistributedBiFunction<Integer, Integer, String> sqlFn,
+            @Nonnull DistributedFunction<ResultSet, T> mapOutputFn
+    ) {
+        this.connectionSupplier = connectionSupplier;
+        this.statementFn = statementFn;
+        this.sqlFn = sqlFn;
+        this.mapOutputFn = mapOutputFn;
+        setCooperative(false);
+    }
+
+    /**
+     * Private API, use {@link SourceProcessors#readJdbcP}.
+     */
+    public static <T> ProcessorSupplier supplier(
+            @Nonnull DistributedSupplier<Connection> connectionSupplier,
+            @Nonnull DistributedFunction<Connection, Statement> statementFn,
+            @Nonnull DistributedBiFunction<Integer, Integer, String> sqlFn,
+            @Nonnull DistributedFunction<ResultSet, T> mapOutputFn
+    ) {
+        return ProcessorSupplier.of(() -> new ReadJdbcP<>(connectionSupplier, statementFn, sqlFn, mapOutputFn));
+    }
+
+    @Override
+    protected void init(@Nonnull Context context) {
+        connection = connectionSupplier.get();
+        statement = statementFn.apply(connection);
+        sql = sqlFn.apply(context.totalParallelism(), context.globalProcessorIndex());
+    }
+
+    @Override
+    public boolean complete() {
+        if (traverser == null) {
+            ResultSet resultSet = uncheckCall(() -> statement.executeQuery(sql));
+            traverser = ((Traverser<ResultSet>) () -> uncheckCall(() -> resultSet.next() ? resultSet : null))
+                    .map(mapOutputFn);
+        }
+        return emitFromTraverser(traverser);
+    }
+
+    @Override
+    public void close(@Nullable Throwable error) throws Exception {
+        if (statement != null) {
+            statement.close();
```

Comment:
this can throw exception, which must be caught

Reply:
Closing the connection closes all the statements. No need to catch, exception in `close` is just logged and ignored.

Reply:
I meant that if statement.close() throws connection.close() will not be called

Reply:
I know, I meant `connection.close()` is enough, no need to close the statement.

Reply:
Checking google, people tend to recommend closing all three: connection, statement and resultset. Even though after connection is closed there's nothing to do for the other two, they say that it's not specified and that some drivers misbehave...